spring-integration-core/src/main/java/org/springframework/integration/aggregator/ResequencingMessageHandler.java

```diff
@@ -81,13 +83,21 @@ protected void afterRelease(MessageGroup messageGroup, Collection<Message<?>> co
 			remove(messageGroup);
 		}
 		else {
-			if (completedMessages != null){
-				int lastReleasedSequenceNumber = this.findLastReleasedSequenceNumber(messageGroup.getGroupId(), completedMessages);
-				messageStore.setLastReleasedSequenceNumberForGroup(messageGroup.getGroupId(), lastReleasedSequenceNumber);
-				this.messageStore.removeMessagesFromGroup(messageGroup.getGroupId(), completedMessages);
+			Object groupId = messageGroup.getGroupId();
+			if (completedMessages != null) {
+				int lastReleasedSequenceNumber = findLastReleasedSequenceNumber(groupId, completedMessages);
+				this.messageStore.setLastReleasedSequenceNumberForGroup(groupId, lastReleasedSequenceNumber);
+				if (this.messageStore instanceof SimpleMessageStore
+						&& completedMessages.size() == messageGroup.size()) {
+					((SimpleMessageGroup) messageGroup).clear();
+					((SimpleMessageGroup) messageGroup).setLastModified(System.currentTimeMillis());
```

Comment:
To `SimpleMessageStore`...
Yes, that's perfect!
We can add it to the `MGS` contract already only in the `4.3`.

Thank you for the pointer! :smile: